TypeScript Client Usage Documentation (#1709)
commercetools/nodejs: docs/sdk/api/typescriptSdk.md

    /*...*/
  })
```

# TypeScript SDK Client v2

> The **@commercetools/sdk-client-v2** is the TypeScript the package that facilitates HTTP [requests](https://commercetools.github.io/nodejs/sdk/Glossary.html#clientrequest) to the platform, ml or history api by using a predefined set of middlewares.

Reviewer: ```suggestion
The **@commercetools/sdk-client-v2** is the TypeScript package that facilitates HTTP [requests](https://commercetools.github.io/nodejs/sdk/Glossary.html#clientrequest) to the Platform, ML or History API by using a predefined set of middlewares.
```

> Unlike the nodejs client, the TypeScript client is a little different in it's usage, however, they are both very similar in every other way and backwards compatible with the nodejs client.

Reviewer: ```suggestion
Unlike the Node.js client, the TypeScript client is a little different in its usage. However, they are both very similar in every other way and backwards compatible with the Node.js client.
```

### Install

```bash
npm install --save @commercetools/sdk-client-v2
or
yarn add @commercetools/sdk-client-v2
```

<!-- ### Browser
```js
<script src="https://unpkg.com/@commercetools/platform-sdk/dist/platform-sdk.umd.js"></script>
<script>
  // global: TypescriptSdk
</script>
``` -->

### Usasge example

Example on how to build a client using only the `defaultClient` class method. The default client enables client creation using very minimal configuratioin.

> ```js

Reviewer: ```suggestion
```typescript
```

// ClientBuilder.js
import { ClientBuilder, Client } from '@commercetools/sdk-client-v2'

const projectKey = 'demo-key'
const oauthUri = 'https://auth.europe-west1.gcp.commercetools.com'
const baseUri = 'https://api.europe-west1.gcp.commercetools.com'
const credentials = {
  clientId: 'clientID12345',
  clientSecret: 'clientSecret12345',
}

const client: Client = new ClientBuilder()
  .defaultClient(baseUri, credentials, oauthUri, projectKey)
  .build()
```

> It is also important to note that we can also chain other middlewares or further configure the default client to suite the specific need of the client being built.

Reviewer: ```suggestion
It is also important to note that we can also chain other middlewares or further configure the default client to suit the specific need of the client being built.
```

  host: 'https://api.europe-west1.gcp.commercetools.com',
  fetch,
}

export const client: Client = new ClientBuilder()
  .withProjectKey(projectKey) // not necessary if the projectKey was already passed in the authMiddlewareOptions
  .withClientCredentialsFlow(authMiddlewareOptions)
  .withHttpMiddleware(httpMiddlewareOptions)
  .withLoggerMiddleware()
  .build()
```

Now the Client object is ready to be used to make HTTP calls.

Example

```typescript
import { client } from './ClientBuilder'
import {
  ApiRoot,
  createApiBuilderFromCtpClient,
} from '@commercetools/platform-sdk'

const apiRoot: ApiRoot = createApiBuilderFromCtpClient(client)

// to get a project details
const getProject = () => {
  return apiRoot.withProjectKey({ projectKey }).get().execute()
}

getProject().then(console.log).catch(console.error)
```

`-`

# Middlewares

Below is a list of all the class methods that can be invoked to add the functionality of a given middleware.

### The authMiddleware methods

These are class methods that creates auth middlewares using different authentication flows or method.

#### withAnonymousSessionFlow(options: _AnonymousAuthMiddlewareOptions_)

> Creates a [middleware](https://commercetools.github.io/nodejs/sdk/Glossary.html#middleware) to handle authentication for the [Anonymous Session](https://docs.commercetools.com/http-api-authorization.html#tokens-for-anonymous-sessions) Flow of the commercetools platform API.

Reviewer: ```suggestion
Creates a [middleware](https://commercetools.github.io/nodejs/sdk/Glossary.html#middleware) to handle authentication for the [Anonymous Session](https://docs.commercetools.com/http-api-authorization.html#tokens-for-anonymous-sessions) flow of the commercetools platform API.
```

#### Named arguments (options)

1. `host` (_String_): the host of the OAuth API service
2. `projectKey` (_String_): the key of the project to assign the default scope to
3. `credentials` (_Object_): the client credentials for authentication (`clientId`, `clientSecret`, `anonymousId`)
4. `scopes` (_Array_): a list of scopes (default manage_project:{projectKey}) to assign to the OAuth token
5. `fetch` (_Function_): A fetch implementation which can be e.g. node-fetch or unfetch but also the native browser fetch function. Only needs be be passed if not globally available (e.g. through isomorphic-fetch)

#### Usage Example

```typescript
import fetch from 'node-fetch'
import { ClientBuilder, Client, AnonymousAuthMiddlewareOptions } from "@commercetools/sdk-client-v2"

const options: AnonymousAuthMiddlewareOptions = {
  host: 'https://auth.europe-west1.gcp.commercetools.com',
  projectKey: 'demo-key',
  credentials: {
    clientId: 'clientID12345',
    clientSecret: 'clientSecret12345',

>     anonymousId: 'client-anonymouseID',

Reviewer: ```suggestion
    anonymousId: 'client-anonymousID',
```

  },
  scopes: ['manage_project:demo-key'],
  fetch,
}

const client: Client = new ClientBuilder()
  .withAnonymousSessionFlow(options)
  ...
```

`-`

#### withClientCredentialsFlow(options: _AuthMiddlewareOptions_)

> Creates a [middleware](https://commercetools.github.io/nodejs/sdk/Glossary.html#middleware) to handle authentication for the Client [Credentials Flow]() of the commercetools platform API.

Reviewer: Missing link for `Credentials Flow`.

Arguments
`code` (_Number_): the HTTP status code
Returns
(`Error` or `undefined`): A [custom error](https://commercetools.github.io/nodejs/sdk/Glossary.html#httperrortype) type (e.g. `BadRequest`, `Unauthorized`) if the code matches, otherwise `undefined`.

#### Usage example

```typescript
import { getErrorByCode } from '@commercetools/sdk-client-v2'

const ErrorType = getErrorByCode(400)
const error = new ErrorType('Oops')
```

`-`

#### withUserAgentMiddleware(options: _UserAgentMiddlewareOptions_)

Creates a [middleware](https://commercetools.github.io/nodejs/sdk/Glossary.html#middleware) to append the `User-Agent` HTTP header to the request.

#### Usage example

```typescript
import { ClientBuilder, Client } from "@commercetools/sdk-client-v2"

const client: Client = new ClientBuilder()
  .withUserAgentMiddleware()
  ...

// The User-Agent will be something like:
// commercetools-js-sdk Node.js/6.9.0 (darwin; x64) my-awesome-library/1.0.0 (+https://github.com/commercetools/my-awesome-library; +helpdesk@commercetools.com)
```

`-`

#### withCorrelationIdMiddleware(options: _CorrelationIdMiddlewareOptions_)

> Creates a middleware to add a correlation id to executed [requests](https://commercetools.github.io/nodejs/sdk/Glossary.html#clientrequest).

Reviewer: ```suggestion
Creates a middleware to add a correlation ID to execute [requests](https://commercetools.github.io/nodejs/sdk/Glossary.html#clientrequest).
```
